### DISCUSSION TITLE
Integrating javascript third party / inline

### DISCUSSION THREAD

**Discussion by iphigenie** [breezefront/community#32] [4. Show and tell]
It was just different enough from documentation example that I tried it the wrong way 8 times at least so I thought I'd share. Please point out if I overlooked something

Integrating the simple jquery "drizzle" to filter things on page repository: https://github.com/anishnair02/drizzle (see https://anishnair02.github.io/drizzle/ for demo with explanation)

drizzle is a script that after initialisation assigns functions to specific elements - these functions are closures. It took me far too much trial and error, with hindsight .

"pure" drizzle.js has the following structure:
```
    try{
        !function(n){
            n.fn.drizzle=function(t){ 
            CODE 
     }}(jQuery)} catch(n) {console.log("Caught an error",n)}
```

magento2 breezed modification - this also works on non breeze :)

```
define(['jquery'], function(n) {
    "use strict";
    
    var result = function(t) {
              CODE 
    }
    if ($.breezemap) {
            $.breezemap['js/drizzle'] = result;
    }
    n.fn.drizzle = result;
    return result;    
});
```

This is then registered in the breeze_default.xml as per documentation and used in templates:

```
<script data-breeze>
require([ 'jquery', 'drizzle'], function($) {
    var filterinstance = $('.filterme-list').drizzle({'child':'.filterme'});
    <?php  foreach ($_optionCollection as $_option): ?>
         < ... >
         $('#<?php echo $_optionTitle ?>').on('click', function(){
            filterinstance.filter('.<?php echo $_optionTitle ?>');
        });
    <?php endforeach; ?>
    
    $('#reset').on('click', function(){
        filterinstance.rain();
    });

});
</script>
```

This seems to work both on my dev site which has breeze enabled and my staging site which hasn't, though it's not been tested very deeply. Not having to have too many duplicate versions while migrating is certainly good

**Comment by vovayatsyuk** [comment]
Wow! That's awesome that you've made it to work with both Luma and Breeze frontends!

This part is not needed for breeze in this case because you use drizzle from the global `$.fn` namespace: 

```
    if ($.breezemap) {
            $.breezemap['js/drizzle'] = result;
    }
```

I also recommend you check if everything works correctly when you navigate to this page from another one, and navigate using back/forward buttons. These are the common cases that require adding destroy listeners to work well with turbo.js which is the part of breeze.

Usually, you'll need to add something like this:

```
$(document).on('breeze:destroy', function () {
    // remove your event listeners from window and body
    // maybe you'll also need to remove js generated markup
});
```

Or, when using a component you can use `destroy` method (breeze will call it automatically):

https://github.com/breezefront/module-breeze/blob/master/view/frontend/web/js/components/messages.js#L85-L89
https://github.com/breezefront/module-breeze/blob/master/view/frontend/web/js/components/swatch-renderer.js#L258-L261

**Comment by iphigenie** [comment]
Thanks, I'll clean it upand test some more

Am I correct in understanding that when there is a

define(['jquery'], function(

jquery is not added or loaded and instead the $ is sent that, because there's cash, methods should work

The challenge is going to be for the third party modules, how to wrap around their (lazily automatically with jquery) javascript in the simplest way possible. Especially the ones using the jQuery variable a lot. 

There's not that many on my site but it's still going to be a shlep.

**Reply by vovayatsyuk** [reply]
We have `require` and `define` functions that know some of the dependencies: https://github.com/breezefront/module-breeze/blob/master/view/frontend/web/js/core/define.js#L4-L13

If some module uses `jQuery` variable, you can try to map it to cash: `jQuery = cash`. But the cash is not completely compatible with jQuery API 🤷‍♂️

**Reply by iphigenie** [reply]
understood - but most places use jquery when they barely need it, so in most cases it'll be enough

**Reply by iphigenie** [reply]
@vovayatsyuk this is now broken by recent changes - I had not moved forward with rebuilding my site on breeze and now that I return to it is seems this no longer works... 

`<script data-breeze>
require([ 'jquery', 'drizzle'], function($,drizzle) {
        var thefilterinstance = $('.filterme-list').drizzle({'child':'.filterme'});`

now results in ".drizzle is not a function"

I suspect it's the same change that #325 covers but returning after a while I can't quite put my finger on what needs to be changed
